File: Kattis/gottatradesomeofem.hs
import           Control.Arrow         ((>>>))
import           Control.Monad         (foldM)
import           Control.Monad.ST      (ST, runST)
import           Data.Array            (Array, (!))
import           Data.Array.ST         (Ix, MArray (..), STUArray, getElems,
                                        readArray, runSTArray, writeArray)
import           Data.Bool             (bool)
import qualified Data.ByteString.Char8 as C
import           Data.Function         ((&))
import           Data.Functor          ((<&>))
import           Data.Maybe            (fromJust)

main :: IO ()
main = do
    [n,m,k]:xs <- C.getContents <&> (C.lines >>> map (C.words >>> map readInt))

    let graph = runSTArray $ do
            g <- newArray (1,n) []
            mapM_ (\[a,b] -> modifyArray g a (b:) >> modifyArray g b (a:)) xs
            pure g

    distribute n k graph
        & maybe "impossible" (map show >>> unwords)
        & putStrLn

distribute :: Int -> Int -> Array Int [Int] -> Maybe [Int]
distribute n k graph = runST $ do
    games <- newArray (1,n) 0 :: ST s (STUArray s Int Int)
    valid <- mapM (\u -> readArray games u >>= ((>0) >>> bool (dfs games u 1 <&> (>k)) (pure True))) [1..n]
    if and valid
        then getElems games <&> Just
        else pure Nothing
  where
    dfs :: STUArray s Int Int -> Int -> Int -> ST s Int
    dfs games u g = do
        writeArray games u (min k g)
        foldM (\r v -> readArray games v >>= ((>0) >>> bool (dfs games v r) (pure r))) (g+1) (graph ! u)

modifyArray :: (MArray a b m, Ix i) => a i b -> i -> (b -> b) -> m ()
modifyArray arr ix f = readArray arr ix >>= (f >>> writeArray arr ix)

readInt :: C.ByteString -> Int
readInt = C.readInt >>> fromJust >>> fst

File: Kattis/gottatradesomeofem.py
n, m, k = (int(i) for i in input().split())

graph = [[] for _ in range(n)]
for _ in range(m):
    u, v = (int(i)-1 for i in input().split())
    graph[u].append(v)
    graph[v].append(u)

games = [None] * n

def search(s):
    g = 1
    curr = [s]
    games[s] = g
    while len(curr):
        next = []
        for u in curr:
            for v in graph[u]:
                if games[v] is None:
                    games[v] = min(k,(g+1))
                    next.append(v)
                    g += 1
        curr = next
    return g >= k

for u in range(n):
    if games[u] is None:
        if not search(u):
            print('impossible')
            exit()

print(*games)
